kkrings/chapter03/exercise_3.11.8.cpp

```diff
@@ -0,0 +1,359 @@
+         * @return Show last element.
+         */
+        T& top()
+        {
+            if (empty())
+            {
+                throw std::underflow_error("The stack is empty.");
+            }
+
+            return data_[size_ - 1];
+        }
+
+        /**
+         * Remove last element.
+         */
+        void pop()
+        {
+            if (empty())
+            {
+                throw std::underflow_error("The stack is empty.");
+            }
+
+            --size_;
+        }
+
+        /**
+         * Insert new element.
+         * @param element New element
+         */
+        void push(const T& element)
+        {
+            if (full())
+            {
+                throw std::overflow_error("The stack is full.");
+            }
+
+            data_[size_] = element;
+            ++size_;
+        }
+
+        /**
+         * Delete all entries;
+         */
+        void clear()
+        {
+            size_ = 0;
+        }
+
+        /**
+         * @return Number of elements
+         */
+        unsigned int size() const
+        {
+            return size_;
+        }
+
+        /**
+         * @return Check whether stack is full.
+         */
+        bool full() const
+        {
+            return size() == max_size;
+        }
+
+        /**
+         * @return Check whether stack is empty.
+         */
+        bool empty() const
+        {
+            return size() == 0;
+        }
+
+    private:
+        std::unique_ptr<T[]> data_;
+        unsigned int size_;
+};
+
+/**
+ * @brief Specialized stack implementation for booleans
+ *
+ * Save 8 booleans into one byte based on char.
+ */
+template<unsigned int max_size>
+class stack<bool, max_size>
+{
+    public:
+        stack() : size_(0), data_(new unsigned char[real_max_size()]) {}
+
+        stack(const stack& other)
+            : size_(other.size()), data_(new unsigned char[real_max_size()])
+        {
+            std::copy(
+                &other.data_[0], &other.data_[real_max_size()], &data_[0]);
+        }
+
+        stack(stack&&) = default;
+
+        ~stack() = default;
+
+        stack& operator=(const stack& other)
+        {
+            assert(other.real_max_size() == real_max_size());
+
+            std::copy(
+                &other.data_[0], &other.data_[real_max_size()], &data_[0]);
+
+            return *this;
+        }
+
+        stack& operator=(stack&&) = default;
+
+        bool top() const
+        {
+            if (empty())
+            {
+                throw std::underflow_error("The stack is empty.");
+            }
+
+            return (data_[(size_ - 1) / 8] >> (size_ - 1) % 8) & 1;
+        }
+
+        void pop()
+        {
+            if (empty())
+            {
+                throw std::underflow_error("The stack is empty.");
+            }
+
+            --size_;
+        }
+
+        void push(bool element)
+        {
+            // unset last bit
+            data_[size_ / 8] &= ~(1 << size_ % 8);
```

Comment:
If I want to push to a stack I have popped from before, which simply decrements the index, I have to make sure that the old value is false.